src/main/kotlin/org/rust/ide/inspections/RsDeprecationInspection.kt

```diff
@@ -0,0 +1,94 @@
+/*
+ * Use of this source code is governed by the MIT license that can be
+ * found in the LICENSE file.
+ */
+
+package org.rust.ide.inspections
+
+import com.intellij.codeInspection.ProblemHighlightType.LIKE_DEPRECATED
+import com.intellij.codeInspection.ProblemsHolder
+import com.intellij.psi.PsiElement
+import com.intellij.psi.PsiElementVisitor
+import com.intellij.psi.util.parents
+import org.rust.lang.core.psi.RsFile
+import org.rust.lang.core.psi.RsMetaItem
+import org.rust.lang.core.psi.RsModDeclItem
+import org.rust.lang.core.psi.RsVisitor
+import org.rust.lang.core.psi.ext.*
+
+class RsDeprecationInspection : RsLocalInspectionTool() {
+    override fun getDisplayName() = "Deprecated item"
+
+    override fun buildVisitor(holder: ProblemsHolder, isOnTheFly: Boolean): PsiElementVisitor = object : RsVisitor() {
+        override fun visitElement(ref: RsElement) {
+            // item is non-inline module declaration or not reference element
+            if (ref is RsModDeclItem || ref !is RsWeakReferenceElement) return
+
+            val original = ref.reference?.resolve() ?: return
+            val identifier = ref.referenceNameElement ?: return
+
+            val targetElement = when {
+                original is RsFile -> original.declaration
+                original is RsAbstractable && original.owner.isImplOrTrait -> original.superItem ?: original
+                else -> original
+            } ?: return
+
+            if (ref.parents().none { it.hasAllowDeprecatedAttribute() }) {
+                checkAndRegisterAsDeprecated(identifier, targetElement, holder)
+            }
+        }
+    }
+
+    private fun checkAndRegisterAsDeprecated(identifier: PsiElement, original: PsiElement, holder: ProblemsHolder) {
+        if (original is RsOuterAttributeOwner) {
+            QueryAttributes(original)
+                .deprecatedAttribute
+                ?.let { DeprecatedAttribute.from(it) }
+                ?.let { holder.registerProblem(identifier, it.getMessage(identifier.text), LIKE_DEPRECATED) }
+        }
+    }
+
+    private fun PsiElement.hasAllowDeprecatedAttribute(): Boolean = this is RsDocAndAttributeOwner &&
+        QueryAttributes(this).hasAttributeWithArg(ALLOW_ATTR_NAME, DEPRECATED_ARG_NAME)
+
+    companion object {
+        private const val ALLOW_ATTR_NAME: String = "allow"
+        private const val DEPRECATED_ARG_NAME: String = "deprecated"
+    }
+}
+
+private class DeprecatedAttribute(val note: String?, val since: String?) {
```

Comment:
Why do we need this class at all?
Looks like we can just create a private function for message generation